pkg/api/http_server.go

```diff
@@ -841,3 +778,146 @@ func (hs *HTTPServer) getDefaultCiphers(tlsVersion uint16, protocol string) []ui
 	}
 	return nil
 }
+
+func (hs *HTTPServer) readCertificates() (*tls.Certificate, error) {
+	if hs.Cfg.CertFile == "" {
+		return nil, errors.New("cert_file cannot be empty when using HTTPS")
+	}
+
+	if hs.Cfg.KeyFile == "" {
+		return nil, errors.New("cert_key cannot be empty when using HTTPS")
+	}
+
+	if _, err := os.Stat(hs.Cfg.CertFile); os.IsNotExist(err) {
+		return nil, fmt.Errorf(`cannot find SSL cert_file at %q`, hs.Cfg.CertFile)
+	}
+
+	if _, err := os.Stat(hs.Cfg.KeyFile); os.IsNotExist(err) {
+		return nil, fmt.Errorf(`cannot find SSL key_file at %q`, hs.Cfg.KeyFile)
+	}
+
+	tlsCert, err := tls.LoadX509KeyPair(hs.Cfg.CertFile, hs.Cfg.KeyFile)
+	if err != nil {
+		return nil, fmt.Errorf("could not load SSL certificate: %w", err)
+	}
+	return &tlsCert, nil
+}
+
+func (hs *HTTPServer) configureTLS() error {
+	tlsCerts, err := hs.tlsCertificates()
+	if err != nil {
+		return err
+	}
+
+	minTlsVersion, err := util.TlsNameToVersion(hs.Cfg.MinTLSVersion)
+	if err != nil {
+		return err
+	}
+
+	tlsCiphers := hs.getDefaultCiphers(minTlsVersion, string(hs.Cfg.Protocol))
+
+	hs.log.Info("HTTP Server TLS settings", "scheme", hs.Cfg.Protocol, "Min TLS Version", hs.Cfg.MinTLSVersion,
+		"configured ciphers", util.TlsCipherIdsToString(tlsCiphers))
+
+	tlsCfg := &tls.Config{
+		Certificates: tlsCerts,
+		MinVersion:   minTlsVersion,
+		CipherSuites: tlsCiphers,
+	}
+
+	hs.httpSrv.TLSConfig = tlsCfg
+
+	if hs.Cfg.Protocol == setting.HTTP2Scheme {
+		hs.httpSrv.TLSConfig.NextProtos = []string{"h2", "http/1.1"}
+	}
+
+	if hs.Cfg.Protocol == setting.HTTPSScheme {
+		hs.httpSrv.TLSNextProto = make(map[string]func(*http.Server, *tls.Conn, http.Handler))
+	}
+
+	if hs.Cfg.CertFile != "" && hs.Cfg.KeyFile != "" {
+		if hs.Cfg.CertWatchInterval != 0 {
+			hs.httpSrv.TLSConfig.GetCertificate = hs.GetCertificate
+			go hs.WatchAndUpdateCerts()
+			hs.log.Info("HTTP Server certificates reload feature is enabled")
+		} else {
+			hs.log.Info("HTTP Server certificates reload feature is NOT enabled")
+		}
+	}
+
+	return nil
+}
+
+func (hs *HTTPServer) GetCertificate(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+	hs.tlsCerts.certLock.RLock()
+	tlsCerts := hs.tlsCerts.certs
+	hs.tlsCerts.certLock.RUnlock()
```

Comment:
Might be clearer to use defer here

```suggestion
	defer hs.tlsCerts.certLock.RUnlock()
	tlsCerts := hs.tlsCerts.certs
```